File: LeetCode/evaluate_reverse_polish_notation.cpp
class Solution {
public:
    bool isNumber(string str) {
        if(isdigit(str[0]))
            return true;

        if(str[0] == '-' && str.size() > 1)
            return true;
        
        return false;
    }

    int evalRPN(vector<string>& tokens) {
        stack<long long int> st;

        for(string token : tokens) {
            if(isNumber(token))
                st.push(stoi(token));
            else {
                int b = st.top();
                st.pop();

                int a = st.top();
                st.pop();

                if(token == "+")
                    st.push(a + b);
                else if(token == "-")
                    st.push(a - b);
                else if(token == "*")
                    st.push((long long) a * b);
                else
                    st.push(a / b);
            }
        }

        return st.top();
    }
};

File: LeetCode/evaluate_reverse_polish_notation.go
type Stack struct {
    arr []int
}

func (s *Stack) push(a int) {
    s.arr = append(s.arr, a)
}

func (s *Stack) pop() int {
    size := len(s.arr)
    tmp := s.arr[size-1]
    s.arr = s.arr[:size-1]
    return tmp
}

func evalRPN(tokens []string) int {
    stack := Stack{
        arr: []int{},
    }

    for _, token := range tokens {
        if token == "+" {
            sum := stack.pop() + stack.pop()
            stack.push(sum)
        } else if token == "-" {
            a, b := stack.pop(), stack.pop()
            sub := b - a
            stack.push(sub)
        } else if token == "*" {
            mult := stack.pop() * stack.pop()
            stack.push(mult)
        } else if token == "/" {
            a, b := stack.pop(), stack.pop()
            div := b / a
            stack.push(div)
        } else {
            number, _ := strconv.Atoi(token)
            stack.push(number)
        }
    }

    return stack.pop() 
}
